drivers: net: sam_gmac: update for supporting multi GMAC instances

TonyHan11 opened the pull request:
The purpose of this PR is to update `sam_gmac` driver from supporting only one instance to supporting multiple instances.

The following two files are updated:
- `drivers/ethernet/eth_sam_gmac.c`
- `drivers/ethernet/eth_sam_gmac_priv.h`

Changes include:
- centralize the lists used for different GMAC queues
- add instance and num_queues to configuration
- update for supporting multi GMAC instances

TonyHan11 committed:
drivers: net: sam_gmac: update macros for multi GMAC instances support

Change the driver from support 1 instance to support multi instances.

Changes includes:
 - irq_config()
 - config & data definitions
 - ETH_NET_DEVICE_DT_INST_DEFINE
 - PTP content / init / DEVICE_DEFINE

Limitations:
 - generate_mac() to be updated for supporting multi instances.
 - the configurations in Kconfig.sam_gmac are used for all instances.

Signed-off-by: Tony Han <tony.han@microchip.com>

TonyHan11 committed:
drivers: net: sam_gmac: add var num_queues to config, update LOG info

Changes includes:
 - add the variable for num_queues to eth_sam_dev_cfg
 - update LOG with dev->name to distinguish different instances

Signed-off-by: Tony Han <tony.han@microchip.com>

TonyHan11 committed:
drivers: net: sam_gmac: centralize the lists used for different queues

The lists optimized includes:
 - RX descriptors list
 - TX descriptors list
 - RX buffer accounting list
 - TX buffer accounting list
 - TX frames accounting list

Signed-off-by: Tony Han <tony.han@microchip.com>

TonyHan11 commented on drivers/ethernet/eth_sam_gmac_priv.h:
```diff
@@ -264,6 +264,9 @@ struct gmac_queue {
 
 /* Device constant configuration parameters */
 struct eth_sam_dev_cfg {
+	uint32_t instance:8;
+	uint32_t num_queues:8;
+	uint32_t reserved:16;
```
No special reasons for using bitfields just it's easier to keep the structure word aligned.
Will be updated with uint16_t and uint8_t, thanks.

fabiobaltieri commented:
Hey, this needs a rebase and force push to pick up the CI error workaround.

TonyHan11 committed:
drivers: net: sam_gmac: update PHY connection type for multi instances

Add variable "phy_conn_type" to get and use the phy_connection_type from
DT for different GMAC instances.
Update the judgement on phy_connection_type for multi instances support.

Signed-off-by: Tony Han <tony.han@microchip.com>

TonyHan11 committed:
drivers: net: sam_gmac: remove run once check in eth_iface_init()

To allow every interface be initialized properly when there are more than
one instance, remove the static variable "init_done" which is used to make
the initialize procedure only be done once.

Signed-off-by: Tony Han <tony.han@microchip.com>

TonyHan11 commented:
@maass-hamburg 
Thank you for your comments, updated with the following changes:
- remove bitfield variable `instance` and use `dev->name` for logging
- assign `.mac_addr` using macro `DT_INST_PROP_OR()`
- move `DT_INST_FOREACH_STATUS_OKAY` to the end and use it only once
- remove `init_done` from the initialize procedure
- PHY connection type: add multi instances support, remove build assert check (errors reported at runtime)
- add multi instances support for MAC address

`#define GMAC_QUEUE_NUM DT_INST_PROP(0, num_queues)` in the header file not changed. Currently, we assume that the queue numbers of all instances are the same, otherwise the code and macros related to `GMAC_QUEUE_NUM` need to be updated. The code and macros are also be affected by some parameters come from `CONFIG_`. It's too complicated to put them in one pull request.

maass-hamburg commented on drivers/ethernet/eth_sam_gmac.c:
```diff
@@ -2109,239 +2036,210 @@ static const struct ethernet_api eth_api = {
+				(MAC_ADDR_SOURCE_EEPROM),				\
+		(COND_CODE_1(DT_INST_PROP(n, zephyr_random_mac_address),		\
+				(MAC_ADDR_SOURCE_RANDOM),				\
+		(COND_CODE_1(DT_NODE_HAS_PROP(DT_DRV_INST(n), local_mac_address),	\
+				(MAC_ADDR_SOURCE_LOCAL),				\
+				(MAC_ADDR_SOURCE_INVALID))))))
```
no need to check for local address, also a invalid mac address should not stop the init as the user can change it later via the api.

maass-hamburg commented on drivers/ethernet/eth_sam_gmac.c:
```diff
@@ -1729,12 +1660,10 @@ static int eth_initialize(const struct device *dev)
 
-#if DT_INST_NODE_HAS_PROP(0, mac_eeprom)
-static void get_mac_addr_from_i2c_eeprom(uint8_t mac_addr[6])
+static void get_mac_addr_from_i2c_eeprom(uint8_t mac_addr[6], const struct i2c_dt_spec i2c)
```
is this part documented in the datasheet or other documentation of the soc manufacturer?
If not it should be removed, as it's up to the user to to set the mac later f.e. via the set_config api. 

Also having ETH_SAM_GMAC_MAC_I2C_INT_ADDRESS and ETH_SAM_GMAC_MAC_I2C_INT_ADDRESS_SIZE as a Kconfig is not a good idea, as it can this way only be set for all instances at one, so for every instance another eeprom has to be used.

maass-hamburg replied:
if it's not documented in a datasheet or a document for the soc, it should be removed, we had a similar case here
https://github.com/zephyrproject-rtos/zephyr/pull/87953#discussion_r2022779300

nandojve commented on drivers/ethernet/eth_sam_gmac.c:
```diff
@@ -1660,34 +1660,9 @@ static int eth_initialize(const struct device *dev)
-#if DT_INST_NODE_HAS_PROP(0, mac_eeprom)
-	get_mac_addr_from_i2c_eeprom(mac_addr);
-#elif DT_INST_PROP(0, zephyr_random_mac_address)
+#if DT_INST_PROP(0, zephyr_random_mac_address)
```
Hi @TonyHan11 ,

I want to continue to use the eeprom mac address and I don't want to use an random mac.
What is the alternative ?

maass-hamburg replied:
setting the mac during runtime directly via 'set_config' or the 'net_mgmt' api

nandojve replied:
What sample do that ?

nandojve replied:
> I don't remember a sample, that shows that

It is because we don't have this in Zephyr yet and here the driver do the bootstrap to the user.

I think what you mention is a great candidate to be automatically applied based on configuration, for instance:

(1) local-mac-address from devicetree
(2) Random address
(3) Eeprom address
(4) None

The network management could do this to the user as a bootstrap and then we can remove this from the driver. User still have the power to complete skip when no option is provided.

In my case, I need to test sometimes on many boards and it is unacceptable to me change ip address each time I change a board. Problems could happen if the random MAC is the same between boards.

nandojve replied:
Using the HWINFO do not solve the use case which is implemented in this driver, it is another alternative, and we can not lost this feature. Users may still require the MAC be inside an EEPROM or similar mechanism which, I think, still not covered and may not be covered by HWINFO.

I'm willing to accept all the improvements to introduce the support to multiple interfaces but not to drop a useful feature that is not available in a generic way in Zephyr yet.

I would like to keep, for the moment, the EEPROM solution in this driver and discuss alternatives that we can implement in a generic way to all types of interfaces.

maass-hamburg replied:
only these sources for mac addresses should be in a ethernet driver:

1. local-mac-address from devicetree
2. random address
3. a kind of unique address generated via HWINFO
4. If specified and documented by the manufacturer other sources like special registers or fuses (user programmable or set by the manufacture)

For production devices you are supposed to use the api to set the mac address, if 4 is not available.

Adding other generic methods into the drivers would lead to duplicated code and makes the drivers more messy and less maintainable.

maass-hamburg replied:
Well, then let's deprecate `CONFIG_ETH_SAM_GMAC_MAC_I2C_EEPROM` (add a `select DEPRECATED`) and only make the dt prop `mac-eeprom` work, when there is only one activated instance. ( a build assert with `!(DT_ANY_INST_HAS_PROP_STATUS_OKAY(mac_eeprom) && ( DT_NUM_INST_STATUS_OKAY(DT_DRV_COMPAT) > 1))`)

TonyHan11 replied:
Updated with the following:
- revert the code related with `mac-eeprom`
- add deprecated to `mac-eeprom` and `ETH_SAM_GMAC_MAC_I2C_EEPROM`
- add `BUILD_ASSERT` to limited `mac-eeprom` be used when there's only one activated GMAC instance

nandojve replied:
Hi @TonyHan11 ,

Could you check #96379
And provide your feedback in #96598

TonyHan11 committed:
drivers: net: sam_gmac: deprecate the 'mac-eeprom' option

Deprecate the 'ETH_SAM_GMAC_MAC_I2C_EEPROM' for the 'mac-eeprom' option,
Limite it to be used when there's only one activated GMAC instance.

Signed-off-by: Tony Han <tony.han@microchip.com>

nandojve commented on drivers/ethernet/eth_sam_gmac.c:
```diff
@@ -174,7 +174,6 @@ BUILD_ASSERT(DT_INST_ENUM_IDX(0, phy_connection_type) <= 1, "Invalid PHY connect
 		DEFN_RX_FRAG(n)								\
 		DEFN_TX_FRAG(n)								\
 		DEFN_TX_FRAME(n)
-DT_INST_FOREACH_STATUS_OKAY(SAM_GMAC_LISTS_DEFN)
```
Be careful with size of commit brief, see [pull-request-guidelines](https://docs.zephyrproject.org/latest/contribute/guidelines.html#pull-request-guidelines)

nandojve commented on drivers/ethernet/eth_sam_gmac.c:
```diff
@@ -133,8 +133,6 @@ static inline void dcache_clean(uint32_t addr, uint32_t size)
 #endif
 #endif /* !CONFIG_NET_TEST */
 
-BUILD_ASSERT(DT_INST_ENUM_IDX(0, phy_connection_type) <= 1, "Invalid PHY connection");
```
ditto

nandojve commented on drivers/ethernet/eth_sam_gmac.c:
```diff
@@ -1660,12 +1660,11 @@ static int eth_initialize(const struct device *dev)
 	return retval;
 }
 
-#if DT_INST_NODE_HAS_PROP(0, mac_eeprom)
```
ditto, pull-request-guidelines

TonyHan11 commented:
Updated with shortening the long commit messages, thanks a lot.

maass-hamburg commented on drivers/ethernet/eth_sam_gmac.c:
```diff
@@ -1752,12 +1686,17 @@ static void get_mac_addr_from_i2c_eeprom(uint8_t mac_addr[6])
-	gen_random_mac(mac_addr, ATMEL_OUI_B0, ATMEL_OUI_B1, ATMEL_OUI_B2);
+#if DT_ANY_INST_HAS_PROP_STATUS_OKAY(mac_eeprom)
+	if (cfg->mac_addr_src == MAC_ADDR_SOURCE_EEPROM) {
+		get_mac_addr_from_i2c_eeprom(mac_addr, cfg->mac_eeprom);
```
we only support this method if only one instance is used, therefore we don't need to save the `const struct i2c_dt_spec` in config and can still use the old code for that.

maass-hamburg reviewed (changes requested):
zephyr_random_mac_address should still be used for every instance, so it needs to go into config

TonyHan11 committed:
drivers: net: sam_gmac: update random MAC addr for multi inst support

Add variable 'random_mac_addr' for 'zephyr,random-mac-address' from
device tree. Update generate_mac() to get random MAC address for each
GAMC interface with the 'zephyr,random-mac-address' property.

Signed-off-by: Tony Han <tony.han@microchip.com>

TonyHan11 commented:
@maass-hamburg a commit added for supporting `zephyr,random-mac-address` for each GMAC instance. Thank you very much.

maass-hamburg reviewed (changes requested):
It might also be good to have a board in tree, that has more than one GMAC instance, so it can be tested

TonyHan11 committed:
drivers: net: sam_gmac: remove getting max_frame_size from DT

As jumbo frame size is not supported by the networking subsystem, only
max_frame_size 1518 and 1536 can be used. The Frame size 1536 would
allow for packets with a vlan tag, so enable GMAC_NCFGR_MAXFS when
NET_VLAN is configured.

Signed-off-by: Tony Han <tony.han@microchip.com>

TonyHan11 committed:
drivers: net: sam_gmac: add BUILD_ASSERT for checking GMAC_QUEUE_NUM

'GMAC_QUEUE_NUM' is a value of 'num-queues' for the first GMAC
instance getting from the device tree. It is used directly or
indirectly (by GMAC_PRIORITY_QUEUE_NUM) for defining and
initializing 'struct eth_sam_dev_data' with a value from Kconfig
(GMAC_ACTIVE_PRIORITY_QUEUE_NUM).

As there will be a big change for applying the corresponding
num-queues for each GMAC 'struct eth_sam_dev_data', here just
keep it as is. Adding the BUITD_ASSERT to make sure the array
queue_list[] is large enough for all GMAC instances.

Signed-off-by: Tony Han <tony.han@microchip.com>

TonyHan11 commented:
@maass-hamburg Many thanks for your comments.

As there's only one GMAC on each SOC currently supported, no more instances is added. For the next supported SOC (SAMA7G5), it has 2 GMACs. The support for SAMA7G5 GMAC will be added based on this PR.

Updated with the following:
- add multi-instance support for `max-frame-size` property
- `COND_CODE_1(DT_INST_PROP(n, zephyr_random_mac_address), (true), (false))`: unuse `COND_CODE_1`
- `#define GMAC_QUEUE_NUM DT_INST_PROP(0, num_queues)`
As there will be a big change for applying the corresponding num-queues for each GMAC `struct eth_sam_dev_data`, here just keep it as is. Adding a BUITD_ASSERT in the last commit to make sure the array queue_list[] is large enough for all GMAC instances.
`GMAC_QUEUE_NUM` is a value of `num-queues` for the first GMAC instance getting from the device tree. It is used directly or indirectly (by GMAC_PRIORITY_QUEUE_NUM) for defining and initializing `struct eth_sam_dev_data` with a value from Kconfig (GMAC_ACTIVE_PRIORITY_QUEUE_NUM).

The same as `GMAC_QUEUE_NUM`, there are many configurations in [Kconfig.sam_gmac](https://github.com/zephyrproject-rtos/zephyr/blob/v4.1.0/drivers/ethernet/Kconfig.sam_gmac#L27) which are used for all the GMAC instances. Might it is better to adapt them for multi-instance support or move them to the device tree in the future.

> can probably use something like `DT_INST_FOREACH_PROP_ELEM()` to iterate over the `interrupts` or `interrupt-names`. (Also use `DT_INST_IRQN_BY_IDX()`) Also looks like `num-queues` is just the amount of interrupts, if thats the case num-queues` is redundant and should be removed.

Same as the above, conditional code controlled by both variables from DT and Kconfig, not easy to use macros to simplify it.

maass-hamburg commented:
@TonyHan11 how about you bring in (limited) support for SAMA7G5 with only one Ethernet controller in the device tree in a separate PR and then add the second controller in this PR. You probably already have this local, as you have to test this code, if it really works for multiple instances.

TonyHan11 commented:
> @TonyHan11 how about you bring in (limited) support for SAMA7G5 with only one Ethernet controller in the device tree in a separate PR and then add the second controller in this PR. You probably already have this local, as you have to test this code, if it really works for multiple instances.

The GMAC support for SAMA7G5 on local depends on the GPIO drivers. I'll create the PR after GPIO is ready.

TonyHan11 commented:
> Hi @TonyHan11 ,
> 
> > > @TonyHan11 how about you bring in (limited) support for SAMA7G5 with only one Ethernet controller in the device tree in a separate PR and then add the second controller in this PR. You probably already have this local, as you have to test this code, if it really works for multiple instances.
> > 
> > 
> > The GMAC support for SAMA7G5 on local depends on the GPIO drivers. I'll create the PR after GPIO is ready.
> 
> DId you create the PR ? If yes, could you CC the reference ?

PR https://github.com/zephyrproject-rtos/zephyr/pull/95135 is created for supporting SAMA7G5 GMAC1, thanks.

nandojve commented:
needs rebase

nandojve commented:
ping @TonyHan11

maass-hamburg commented on drivers/ethernet/eth_sam_gmac.c:
```diff
@@ -1860,17 +1792,31 @@ static void eth0_iface_init(struct net_if *iface)
+		gmac_ncfgr_val |= GMAC_NCFGR_MAXFS;
+		break;
+	case 10240:
+		gmac_ncfgr_val |= GMAC_NCFGR_JFRAME;
```
jumbo frame size is not supported by the networking subsystem, so we shouldn't allow this

TonyHan11 replied:
Added the following to check the frame size at the build stage, thanks.
```
BUILD_ASSERT(DT_INST_PROP(n, max_frame_size) != 10240,			\
	     "jumbo frame size not supported by the networking "	\
	     "sybsystem, fix it at device tree.");					\
```

TonyHan11 committed:
drivers: net: sam_gmac: ref_clk_source for multi instances support

Add variable "ref_clk_source" to get and set the source for the GMAC
reference clock from DT for different GMAC instances.

Signed-off-by: Tony Han <tony.han@microchip.com>

TonyHan11 commented on drivers/ethernet/eth_sam_gmac_priv.h:
```diff
@@ -47,44 +47,48 @@ BUILD_ASSERT(ARRAY_SIZE(GMAC->GMAC_TBQBAPQ) + 1 == GMAC_QUEUE_NUM,
-#define PRIORITY_QUEUE1_RX_DESC_COUNT   1
-#define PRIORITY_QUEUE1_TX_DESC_COUNT   1
-#endif
+#define PRIORITY_QUEUE_RX_DESC_COUNT						\
```
> line 34:
> 
> ```
> #define GMAC_QUEUE_NUM DT_INST_PROP(0, num_queues)
> ```
> 
> still depends on the first instance

> still relevant, now line 30

Yes, the correlation needs to be solved. `GMAC_QUEUE_NUM` is used as the array sizeof of `queue_list` in `struct eth_sam_dev_data`. It comes from the num_queues of the first GMAC instance in DT.

The same `num_queues` is also used in [`Kconfig.sam_gmac`](https://github.com/zephyrproject-rtos/zephyr/blob/v4.3.0-rc2/drivers/ethernet/Kconfig.sam_gmac#L24) for some configurations (also be used for SAMX7X, SAM4E and SAME54 series MCUs). Re-organize the configurations into the DT is a bit complicated and it's too big to put in a same PR with about 10 commits.

In this PR just keep it as is and validate it with `BUILD_ASSERT` in the last commit. It maybe solved in an additional pull request. Thank you very much for your comment.

maass-hamburg commented on drivers/ethernet/eth_sam_gmac.c:
```diff
@@ -1860,17 +1792,31 @@ static void eth0_iface_init(struct net_if *iface)
+		break;
+	case 1536:
+		gmac_ncfgr_val |= GMAC_NCFGR_MAXFS;
+		break;
```
the difference between these two is that 1536 would allow for packets with a vlan tag, right? If so, it should be chosen based on `CONFIG_NET_VLAN` and not the dt prop, so that prop could be removed.

maass-hamburg replied:
you can also remove it from the dt binding yaml

maass-hamburg replied:
remove it directly.

TonyHan11 committed:
dts: bindings: ethernet-controller: remove unused prop 'max-speed'

The 'max-speed' property in atmel,gmac-common.yaml file is not used,
remove  it.

Signed-off-by: Tony Han <tony.han@microchip.com>

TonyHan11 commented on dts/bindings/ethernet/atmel,gmac-common.yaml:
```diff
@@ -42,7 +26,7 @@ properties:
 
   mac-eeprom:
     type: phandle
```
> maybe better to keep this to not break it for existing users. They will already get a warning from the deprecated Kconfig

OK, updated.

TonyHan11 commented on dts/bindings/ethernet/atmel,gmac-common.yaml:
```diff
@@ -16,22 +16,6 @@ properties:
-      to 32 bytes.  The hardware have support to jumbo frames and it can be
-      enabled by selecting the value 10240.
-
   max-speed:
```
`max-speed` removed.

nandojve commented:
Ping @jukkar
